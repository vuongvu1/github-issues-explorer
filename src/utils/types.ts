export type IssueType = Array<{
  id: string;
  title: string;
  number: number;
  createdAt: string;
  author: {
    login: string;
  };
  comments: {
    totalCount: number;
  };
}>;
