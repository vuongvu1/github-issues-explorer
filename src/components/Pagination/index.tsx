import React from "react";
import SC from "./styles";

interface Props {
  current: number;
  hasPreviousPage?: boolean;
  onClickPrevious: () => void;
  hasNextPage?: boolean;
  onClickNext: () => void;
}

const Pagination: React.FC<Props> = ({
  current,
  hasPreviousPage,
  onClickPrevious,
  hasNextPage,
  onClickNext,
}) => {
  return (
    <SC.Container>
      <div>
        <SC.Button
          data-testid="prev-btn"
          isDisabled={!hasPreviousPage}
          onClick={hasPreviousPage ? onClickPrevious : undefined}
        >
          Previous
        </SC.Button>
        {hasPreviousPage && (
          <SC.Button onClick={onClickPrevious}>{current - 1}</SC.Button>
        )}
        <SC.Button isActive>{current}</SC.Button>
        {hasNextPage && (
          <SC.Button onClick={onClickNext}>{current + 1}</SC.Button>
        )}
        <SC.Button
          data-testid="next-btn"
          isDisabled={!hasNextPage}
          onClick={hasNextPage ? onClickNext : undefined}
        >
          Next
        </SC.Button>
      </div>
    </SC.Container>
  );
};

export default Pagination;
