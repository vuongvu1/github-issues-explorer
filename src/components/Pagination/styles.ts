import styled, { css } from "styled-components";

interface ButtonProps {
  isDisabled?: boolean;
  isActive?: boolean;
}

const SC = {
  Container: styled.div(
    ({ theme }) => css`
      display: flex;
      justify-content: center;
      align-items: center;
      flex-wrap: wrap;
      margin-top: ${theme.spacing.md};
    `
  ),
  Button: styled.button<ButtonProps>(
    ({ theme, isActive, isDisabled }) => css`
      padding: ${theme.spacing.sm};
      background-color: ${theme.palette.white};
      min-width: 34px;
      height: 34px;
      cursor: pointer;
      margin-left: ${theme.spacing.sm};
      border-radius: ${theme.borderRadius};
      border: 1px solid transparent;
      transition: all 0.3s;

      &:hover {
        border: 1px solid ${theme.palette.dark};
        background-color: ${theme.palette.neutral};
      }

      ${isDisabled &&
      css`
        opacity: 0.5;
        cursor: not-allowed;
        &:hover {
          background-color: ${theme.palette.white};
        }
      `}

      ${isActive &&
      css`
        font-weight: bold;
        color: ${theme.palette.white};
        background-color: ${theme.palette.success};

        &:hover {
          border: 1px solid transparent;
          background-color: ${theme.palette.success};
        }
      `}
    `
  ),
};

export default SC;
