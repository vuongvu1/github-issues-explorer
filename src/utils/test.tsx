import { ReactNode } from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { BrowserRouter } from "react-router-dom";
import { theme } from "src/globalStyles";
import {
  IssueState,
  IssueOrderField,
  OrderDirection,
} from "src/generated/graphql";

const mockStore = configureMockStore();
const store = mockStore({
  repoSlice: {
    url: "https://github.com/reactjs/reactjs.org",
    currentPage: 1,
    searchParams: {
      name: "reactjs.org",
      owner: "reactjs",
      status: IssueState.Open,
      orderBy: IssueOrderField.CreatedAt,
      orderDir: OrderDirection.Desc,
      before: undefined,
      after: undefined,
      first: 10,
      last: undefined,
    },
  },
});

export const customRender = (Component: ReactNode) =>
  render(
    <ThemeProvider theme={theme}>
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Provider store={store}>{Component}</Provider>
        </BrowserRouter>
      </MockedProvider>
    </ThemeProvider>
  );
