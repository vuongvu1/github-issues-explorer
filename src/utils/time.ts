const ONE_YEAR_IN_SECONDS = 31536000;
const ONE_MONTH_IN_SECONDS = 2592000;
const ONE_DAY_IN_SECONDS = 86400;
const ONE_HOUR_IN_SECONDS = 3600;
const ONE_MINUTE_IN_SECONDS = 60;

export const timeSince = (date: Date) => {
  const seconds = Math.floor((new Date().getTime() - date.getTime()) / 1000);

  let interval = seconds / ONE_YEAR_IN_SECONDS;

  if (interval > 1) {
    return `${Math.floor(interval)} years`;
  }

  interval = seconds / ONE_MONTH_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} months`;
  }

  interval = seconds / ONE_DAY_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} days`;
  }

  interval = seconds / ONE_HOUR_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} hours`;
  }

  interval = seconds / ONE_MINUTE_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} minutes`;
  }

  return `${Math.floor(seconds)} seconds`;
};
