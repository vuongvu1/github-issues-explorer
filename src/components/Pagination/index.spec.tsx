import { cleanup, fireEvent } from "@testing-library/react";
import { customRender } from "src/utils/testUtils";
import Pagination from "./index";

describe("Pagination", () => {
  afterEach(cleanup);

  test("should match snapshot", () => {
    const { container } = customRender(
      <Pagination
        current={1}
        hasPreviousPage={true}
        hasNextPage={false}
        onClickPrevious={jest.fn()}
        onClickNext={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should trigger onClickPrevious and onClickNext", () => {
    const onClickPrev = jest.fn();
    const onClickNext = jest.fn();

    const { getByTestId } = customRender(
      <Pagination
        current={2}
        hasPreviousPage={true}
        hasNextPage={true}
        onClickPrevious={onClickPrev}
        onClickNext={onClickNext}
      />
    );

    const prevBtn = getByTestId("prev-btn");
    fireEvent.click(prevBtn);
    expect(onClickPrev).toHaveBeenCalledTimes(1);

    const nextBtn = getByTestId("next-btn");
    fireEvent.click(nextBtn);
    expect(onClickNext).toHaveBeenCalledTimes(1);
  });
});
