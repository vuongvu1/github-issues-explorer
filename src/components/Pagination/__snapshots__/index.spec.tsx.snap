// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvaa ePfWPJ"
  >
    <div>
      <button
        class="sc-gsDJrp inDnEA"
        data-testid="prev-btn"
      >
        Previous
      </button>
      <button
        class="sc-gsDJrp inDnEA"
      >
        0
      </button>
      <button
        class="sc-gsDJrp gTcmdc"
      >
        1
      </button>
      <button
        class="sc-gsDJrp eMQOXM"
        data-testid="next-btn"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
